### PR TITLE
fix md option and implement geometry opt in arted side

### PR DESCRIPTION
**Pull request by ayamada224**
(1) Bug fix for md option (use_ehrenfest_md) --- allocation & update of zJxyz and zKxyz etc
(2) Implement geometry optimization -- add optimization subroutine, some changes in ground state parts (set convergence threshold, print level, initial guess etc only for optimization). Also added trajectory output files for md and opt. 
(Both are in arted side)

### CODE REVIEW

**Comment by uemoto1** [comment]
add to whitelist

**Comment by uemoto1** [comment]
The above comment is the forcing execution command to Jenkins

**Comment by shunsuke-sato** [comment]
@ayamada224 san, thank you for your implementation. 
- I confirmed that this patch passes the regression tests in the `example` directory. I also tested the geometry optimization for Si from a displaced initial atomic positions, and the code produced a reasonable result.

- Although I had a quick review, I could not carefully review this patch, and it would be a bit difficult to manage the time this week. So, I approve this patch, but would like to ask @uemoto1 san to review this patch closely.

- Finally, I found a minor bug in 'write_xyz' routine, where all the MPI processes write down the data in the same file, and the data were broken. I have fixed the routine and sent a pull-request to your branch. So, please merge it.

Thanks for your contribution.

**Review by shunsuke-sato — APPROVED** [review]

**Comment by uemoto1** [comment]
Thank you for developing and reviewing, @ayamada224 san and @shunsuke-sato, and sorry for late reply. I will review this modification later.

File: src/ARTED/control/ground_state.f90 [hunk 1]
```diff
@@ -135,10 +141,22 @@ subroutine calc_ground_state
           end if
           write(*,*) '-----------------------------------------------'
        end if
+
+       !(exit if energy difference is below threshold: only if set convrg_scf_Eall>0)
+       if( abs(Eall_GS(iter)-Eall_GS(iter-1)) .le. convrg_scf_ene ) then
+          if(kbTev >= 0d0) write(*,*) "sorry, occ.out was not generated" !(fix if need)
```
Comment (uemoto1) [inline src/ARTED/control/ground_state.f90]: The condition of `comm_is_root(nproc_id_global)` will be required here.

File: src/ARTED/control/control_sc.f90 [hunk 2]
```diff
@@ -426,4 +443,499 @@ subroutine reset_rt_timer
...
+  call cal_mean_max_forces(NI,force,fave,fmax)
+  if(comm_is_root(nproc_id_global)) then
+     write(*,135) 0, 0, Eall, 0d0
+     write(*,120) " Max-force=", fmax, "  Mean-force=", fave
+  endif
+  !(initial check of convergence: if force is enough small, exit before opt)
+  if(fave .le. fave_conv) then
+    if(comm_is_root(nproc_id_global)) &
+    &  write(*,*) " Mean force is enough small: stop calculation"
+    stop
```
Comment (uemoto1) [inline src/ARTED/control/control_sc.f90]: I prefer to call `end_parallel` subroutine before stop the program.

File: src/ARTED/control/ground_state.f90 (hunk 1 above)
Comment (uemoto1) [inline src/ARTED/control/ground_state.f90]: The condition of `comm_is_root(nproc_id_global)` might be required here.

**Review by uemoto1 — APPROVED** [review]
Thank you for developing @ayamada224 san.

I grasped this modification contains the bug fix of `use_ehrenfest_md='y'`, which fix the problem of deallocation of `zJxyz`/`mJxyz` in `not_initial` case. This PR also append the new calculation mode of `use_ehrenfest_md='y'` options for the ARETD part, it adds subroutine of `calc_opt_ground_state`  a few modifications to the `calc_ground_state` in order to call from the `calc_opt_ground_state` in each searching iteration.

Accordings to @ayamada224  san's recent calculation report, this optimization method successfully produces the atomic positions of some molecules. I agree this implementation works well.

I think it is available to merge the developer branch.

**Comment by uemoto1** [comment]
Thank you for the implementation, @ayamada224 san.
I think this branch is available to merge, and I will merge it.

**Pull request merged**